### DISCUSSION TITLE
Google OAuth2 access additional User information

### DISCUSSION THREAD

**Discussion by mo-cmyk** [fastapi-users/fastapi-users#1113] [Q&A]
Hi, I implemented the Google OAuth2 Router as shown in the documentation, and created everything necessary on the Google Cloud Console (ID, Secret, etc.). In my Database, my User Model also requires the first and last name of each user which works excellently with a regular register endpoint. Is there any way to access the Google User's last and first name information in "FastAPI Users"?

For now, I am simply getting Validation Errors as soon as the OAuth endpoint is used, due to no first and last name being provided.
Any Help is greatly appreciated! 

I was able to modify the Google OAuth Client to extract also the first and last name as such:

```python
from typing import Any, Dict, List, Optional, Tuple, cast

import httpx
from typing_extensions import Literal, TypedDict

from httpx_oauth.errors import GetIdEmailError
from httpx_oauth.oauth2 import BaseOAuth2

AUTHORIZE_ENDPOINT = "https://accounts.google.com/o/oauth2/v2/auth"
ACCESS_TOKEN_ENDPOINT = "https://oauth2.googleapis.com/token"
REVOKE_TOKEN_ENDPOINT = "https://accounts.google.com/o/oauth2/revoke"
BASE_SCOPES = [
    "https://www.googleapis.com/auth/userinfo.profile",
    "https://www.googleapis.com/auth/userinfo.email",
]
PROFILE_ENDPOINT = "https://people.googleapis.com/v1/people/me"


class GoogleOAuth2AuthorizeParams(TypedDict, total=False):
    access_type: Literal["online", "offline"]
    include_granted_scopes: bool
    login_hint: str
    prompt: Literal["none", "consent", "select_account"]


class GoogleOAuth2(BaseOAuth2[GoogleOAuth2AuthorizeParams]):
    def __init__(
        self,
        client_id: str,
        client_secret: str,
        scope: Optional[List[str]] = BASE_SCOPES,
        name="google",
    ):
        super().__init__(
            client_id,
            client_secret,
            AUTHORIZE_ENDPOINT,
            ACCESS_TOKEN_ENDPOINT,
            ACCESS_TOKEN_ENDPOINT,
            REVOKE_TOKEN_ENDPOINT,
            name=name,
            base_scopes=scope,
        )

    async def get_id_email(self, token: str) -> Tuple[str, str]:
        async with self.get_httpx_client() as client:
            response = await client.get(
                PROFILE_ENDPOINT,
                params={"personFields": "emailAddresses,names"},
                headers={**self.request_headers, "Authorization": f"Bearer {token}"},
            )

            if response.status_code >= 400:
                raise GetIdEmailError(response.json())

            data = cast(Dict[str, Any], response.json())
            user_id = data["resourceName"]
            user_email = next(
                email["value"]
                for email in data["emailAddresses"]
                if email["metadata"]["primary"]
            )
            first_name, last_name = next(
                (name.get('givenName'), name.get('familyName'))
                for name in data["names"]
                if name["metadata"]["primary"]
            )
            print("##########")
            print(first_name)
            print(last_name)
            print("##########")

            return user_id, user_email

```
With a lot of reverse engineering, it is kind of possible to at least create an account using google utilizing first and last name, but is there any way it is supported by the package without me having a local fork of the package or something ?

**Comment by frankie567** [comment]
IMO, your best bet is to define your own `/callback` route. Here is the one provided by FastAPI Users:

https://github.com/fastapi-users/fastapi-users/blob/baca17d9e029aafb757e128cedbb6782e45bcd89/fastapi_users/router/oauth.py#L100-L147

As you can see, the first step is to call `get_id_email`. If you avoid this and make your own call to the Google API, you'll be able to retrieve all the information you need and going through the rest of the process.

**Reply by mo-cmyk** [reply]
Is that the only time the OAuth client instance is accessed so can I just define my own with a method that accesses more than just email as shown above?

**Reply by frankie567** [reply]
Here, and for the `/authorize` route just above `callback`. So yes, should be safe.

**Reply by mo-cmyk** [reply]
So would I also need to override the authorized route? Probably best to rewrite the whole router in this case right ? And define a custom OAuth Class.

**Reply by frankie567** [reply]
Well, yes, but you can keep it untouched.

**Reply by mo-cmyk** [reply]
```python
user = await user_manager.oauth_associate_callback(
            user,
            oauth_client.name,
            token["access_token"],
            account_id,
            account_email,
            token.get("expires_at"),
            token.get("refresh_token"),
            request,
        )
```
The function of the user manager class also needs to be rewritten to accept the name then right?

**Reply by frankie567** [reply]
Not necessarily, you could very well update the user in your router logic without touching the manager methods.

**Reply by mo-cmyk** [reply]
Where would I do that best and how? I am really sorry if my questions are bad I am just really trying to understand the logic here. So to make it easier to understand let's say I implemented everything in a way that I now have access to the user's name via a google API call inside my OAuth router in the callback method, how would I now store that name for the given user without touching my user_manager?

**Reply by mo-cmyk** [reply]
Edit: 
I got it somewhat working by adding this code to my router:

```python
        try:
            user = await user_manager.oauth_callback(
                oauth_client.name,
                token["access_token"],
                account_id,
                account_email,
                token.get("expires_at"),
                token.get("refresh_token"),
                request,
                associate_by_email=associate_by_email,
            )
            if user.first_name != first_name or user.last_name != user.last_name:
                await user_manager.update(user_update=UserUpdate(first_name=first_name, last_name=last_name),user=user, request=request)
```

the problem is that I needed to a) modify my database Model to make the name not required like this:

```python
class OAuthAccount(BaseOAuthAccount):
    """
    OAuth Account Model
    """
    first_name: str = None
    last_name: str = None


class User(BeanieBaseUser[PydanticObjectId]):
    """
    User Base Model (Instructor)
    """
    oauth_accounts: List[OAuthAccount] = Field(default_factory=list)
    first_name: str = None
    last_name: str = None
```

and also my update/create models to this:
```python
class UserRead(schemas.BaseUser[PydanticObjectId]):
    """
    Class for User Read Operations
    """
    first_name: Optional[str] = None
    last_name: Optional[str] = None


class UserCreate(schemas.BaseUserCreate):
    """
    Class for User Create Operations
    """
    first_name: Optional[str] = None
    last_name: Optional[str] = None


class UserUpdate(schemas.BaseUserUpdate):
    """
    Class for User Edit Operations
    """
    first_name: Optional[str] = None
    last_name: Optional[str] = None
```
Therefore if I register using the regular register endpoint it is not required anymore to enter a name which should still bel required.
Any ideas to fix that?

**Reply by mo-cmyk** [reply]
Maybe we also did talk past each other? @frankie567

**Reply by mo-cmyk** [reply]
Actually, I realized my approach isn't really the right, since my on after register, sends an email, addressing a user with their name which would be missing if I set it after register.

**Comment by caniko** [comment]
Thanks for this! This should be in the docs!

**Comment by saeedmv** [comment]
Hi, I am trying to generate REFRESH TOKEN from Google Cloud Console for my Python code.  Also, I already have client_id and client_secret.

could you help me generate REFRESH TOKEN from Google Cloud Console or any other way?

My 'Jason code' is as 'client_secret.json' as file name:
`{
  "web": {
    "client_id": "my_client_id",
    "client_secret": "my_client_secre",
    "redirect_uris": ["urn:ietf:wg:oauth:2.0:oob", "http://localhost"]
  }
}`
my 'credentials code' is as 'credentials.json' as file name:
`{
  "token": "my_token",
  "refresh_token": "my_refresh_token",
  "token_uri": "https://oauth2.googleapis.com/token",
  "client_id": "my_client_id",
  "client_secret": "my_client_secret",
  "scopes": [
    "https://www.googleapis.com/auth/drive"
  ],
  "expiry": "2022-01-01T00:00:00Z"
}`
my refresh token code is :
`from google.oauth2.credentials import Credentials
import json

# Load the client secret file
client_secret_file = "client_secret.json"

# Load the credentials file
with open("credentials.json", "r") as f:
    credentials_data = json.load(f)

# Create the credentials object
creds = Credentials.from_authorized_user_info(info=credentials_data)

# Print the refresh token
print("Refresh Token:", creds.refresh_token)`

Is my activities fine?

**Comment by dikwickley** [comment]
thanks!

**Comment by dikwickley** [comment]
for anyone out there
```
# TODO: Properly implement this https://github.com/fastapi-users/fastapi-users/blob/master/fastapi_users/router/oauth.py#L101-L154
@auth_router.get("/auth/github/custom-callback", tags=["auth"])
async def custom_github_callback(
    request: Request,
    code: str,
    state: str,
    user_manager: BaseUserManager[models.UP, models.ID] = Depends(fastapi_users.get_user_manager),):
    token = await github_oauth_client.get_access_token(code, redirect_uri="http://localhost:3000/api/auth/callback/github")
    account_id, account_email = await github_oauth_client.get_id_email(
            token["access_token"]
        )

    user = await github_oauth_client.get_profile(token['access_token'])

    # HERE it is!
    print(user)

    try:
        user = await user_manager.oauth_callback(
            github_oauth_client.name,
            token["access_token"],
            account_id,
            account_email,
            token.get("expires_at"),
            token.get("refresh_token"),
            associate_by_email=True,
            is_verified_by_default=True
        )
    except UserAlreadyExists:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=ErrorCode.OAUTH_USER_ALREADY_EXISTS,
        )
    
    if not user.is_active:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=ErrorCode.LOGIN_BAD_CREDENTIALS,
        )

    response = await auth_backend.login(auth_backend.get_strategy(), user)
    await user_manager.on_after_login(user, request, response)

    return response
```

and then in your FE just use the `/custom-callback` (PS import all necessary stuff)
